src/extra/gfx/texture.cpp

```diff
@@ -0,0 +1,170 @@
+        else if (asType == TextureType::SNorm)
+          format.pixelFormat = WGPUTextureFormat_R8Snorm;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_R8Uint;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_R8Sint;
+        break;
+      }
+      break;
+    case 2:
+      switch (componentType) {
+      case ComponentType::Float:
+        if (asType == TextureType::Float)
+          format.pixelFormat = WGPUTextureFormat_RG32Float;
+        break;
+      case ComponentType::Int16:
+        if (asType == TextureType::UInt)
+          format.pixelFormat = WGPUTextureFormat_RG16Uint;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_RG16Sint;
+        break;
+      case ComponentType::Int8:
+        if (asType == TextureType::UNorm)
+          format.pixelFormat = WGPUTextureFormat_RG8Unorm;
+        else if (asType == TextureType::SNorm)
+          format.pixelFormat = WGPUTextureFormat_RG8Snorm;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_RG8Uint;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_RG8Sint;
+        break;
+      }
+      break;
+    case 3:
+      throw formatException("RGB textures not supported");
+    case 4:
+      switch (componentType) {
+      case ComponentType::Float:
+        if (asType == TextureType::Float)
+          format.pixelFormat = WGPUTextureFormat_RGBA32Float;
+        break;
+      case ComponentType::Int16:
+        if (asType == TextureType::UInt)
+          format.pixelFormat = WGPUTextureFormat_RGBA16Uint;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_RGBA16Sint;
+        break;
+      case ComponentType::Int8:
+        if (asType == TextureType::UNorm)
+          format.pixelFormat = WGPUTextureFormat_RGBA8Unorm;
+        else if (asType == TextureType::UNormSRGB)
+          format.pixelFormat = WGPUTextureFormat_RGBA8UnormSrgb;
+        else if (asType == TextureType::SNorm)
+          format.pixelFormat = WGPUTextureFormat_RGBA8Snorm;
+        else if (asType == TextureType::UInt)
+          format.pixelFormat = WGPUTextureFormat_RGBA8Uint;
+        else if (asType == TextureType::Int)
+          format.pixelFormat = WGPUTextureFormat_RGBA8Sint;
+        break;
+      }
+      break;
+    }
+
+    if (format.pixelFormat == WGPUTextureFormat_Undefined)
+      throw TextureFormatException(componentType, asType);
+
+    auto &inputFormat = Texture::getInputFormat(format.pixelFormat);
+    size_t imageSize = inputFormat.pixelSize * image.width * image.height;
+
+    // Copy the data since we can't keep a reference to the image variable
+    ImmutableSharedBuffer isb(image.data, imageSize);
+    texture->init(format, int2(image.width, image.height), SamplerState(), std::move(isb));
```

Comment:
Does this support loading a new image every frame efficiently? Like if I wanna render a video etc.
It should ideally

Reply:
Yes, only thing I couldn't avoid at the moment is copying out from the image since I didn't find a way to keep that memory alive until when I issue the GPU commands that need that memory.

Reply:
It's fine cos it likely can be multithreaded or so